### DISCUSSION TITLE
X11 forward over ssh: is it possible with net=none?

### DISCUSSION THREAD

**Discussion by slowpeek** [netblue30/firejail#4323] [Q&A]
_firejail 0.9.64.4 on ubuntu focal here btw_

I've been trying to run firejail'ed keepassxc over ssh like:
```
ssh -XC 127.0.0.1 firejail --profile=keepassxc /usr/bin/keepassxc
```
It doesn't work. Error message:
```
qt.qpa.xcb: could not connect to display localhost:10.0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even
though it was found.
This application failed to start because no Qt platform plugin could be
initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl,
offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl,
wayland-xcomposite-glx, xcb.
```

I figured out such patch to make it work:
```diff
--- keepassxc.profile   2021-02-08 19:23:11.000000000 +0200
+++ test.profile    2021-06-01 01:46:49.707809088 +0300
@@ -43,7 +43,6 @@
 
 caps.drop all
 machine-id
-net none
 no3d
 nodvd
 nogroups
@@ -53,7 +52,6 @@
 notv
 nou2f
 novideo
-protocol unix,netlink
 seccomp !name_to_handle_at
 seccomp.block-secondary
 shell none
@@ -61,7 +59,7 @@
 
 private-bin keepassxc,keepassxc-cli,keepassxc-proxy
 private-dev
-private-etc alternatives,fonts,ld.so.cache,machine-id
+private-etc alternatives,fonts,ld.so.cache,machine-id,hosts
 private-tmp
 
 dbus-user filter
```
After all I wonder: if it is possible in the firejail universe to have net=none and X11 forwarded over ssh?

**Comment by rusty-snake** [comment]
> ```diff
> -protocol unix,netlink
> ```

You should

```diff
-protocol unix,netlink
+protocol unix,inet,inet6,netlink
```

(allow inet,inet6 in additions instead of allow-all)

> if it is possible in the firejail universe to have net=none and X11 forwarded over ssh?

IDK how the X11-forward works, but it seems to use a TCP socket on localhost. Because `net=none` creates a new (empty) network namespace which has no access to the default loopback/localhost (#108) this does not work.

**Reply by slowpeek** [reply]
Thanks for reply. I'm off to keepassxc-cli over ssh for now.

**Comment by zithro** [comment]
X11 forwarding works by creating a `127.0.0.1:60xx` socket, with xx being the $DISPLAY number used by ssh -X.
So with net=none there's no way to access localhost.

There's a hackish and not-so-secure way to circumvent it : allow host's main X11 to listen on external IP (instead of only localhost), and use a `net=eth0` and `netfilter`ed firejail to only forward and allow access to this listening port.

EDIT: don't do this without proper firewalling rules on your host.
